drivers/video/video_stm32_venc.c

```diff
@@ -0,0 +1,899 @@
+			return -EINVAL;
+		}
+
+		fmt->size = fmt->width * fmt->height / VENC_ESTIMATED_COMPRESSION_RATIO;
+		data->out_fmt = *fmt;
+	}
+
+	return 0;
+}
+
+static int stm32_venc_get_fmt(const struct device *dev, struct video_format *fmt)
+{
+	struct stm32_venc_data *data = dev->data;
+
+	if (fmt->type == VIDEO_BUF_TYPE_INPUT) {
+		*fmt = data->in_fmt;
+	} else {
+		*fmt = data->out_fmt;
+	}
+
+	return 0;
+}
+
+static int encoder_prepare(struct stm32_venc_data *data)
+{
+	H264EncRet ret;
+	H264EncConfig cfg = {0};
+	H264EncPreProcessingCfg preproc_cfg = {0};
+	H264EncRateCtrl ratectrl_cfg = {0};
+	H264EncCodingCtrl codingctrl_cfg = {0};
+
+	data->frame_nb = 0;
+
+	/* set config to 1 reference frame */
+	cfg.refFrameAmount = 1;
+	/* frame rate */
+	cfg.frameRateDenom = 1;
+	cfg.frameRateNum = VENC_DEFAULT_FRAMERATE;
+	/* image resolution */
+	cfg.width = data->out_fmt.width;
+	cfg.height = data->out_fmt.height;
+	/* stream type */
+	cfg.streamType = H264ENC_BYTE_STREAM;
+
+	/* encoding level*/
+	cfg.level = VENC_DEFAULT_LEVEL;
+	cfg.svctLevel = 0;
+	cfg.viewMode = H264ENC_BASE_VIEW_SINGLE_BUFFER;
+
+	ret = H264EncInit(&cfg, &data->encoder);
+	if (ret != H264ENC_OK) {
+		LOG_ERR("H264EncInit error=%d", ret);
+		return -EIO;
+	}
+
+	/* set format conversion for preprocessing */
+	ret = H264EncGetPreProcessing(data->encoder, &preproc_cfg);
+	if (ret != H264ENC_OK) {
+		LOG_ERR("H264EncGetPreProcessing error=%d", ret);
+		return -EIO;
+	}
+	preproc_cfg.inputType = to_h264pixfmt(data->in_fmt.pixelformat);
+	ret = H264EncSetPreProcessing(data->encoder, &preproc_cfg);
+	if (ret != H264ENC_OK) {
+		LOG_ERR("H264EncSetPreProcessing error=%d", ret);
+		return -EIO;
+	}
+
+	/* setup coding ctrl */
+	ret = H264EncGetCodingCtrl(data->encoder, &codingctrl_cfg);
+	if (ret != H264ENC_OK) {
+		LOG_ERR("H264EncGetCodingCtrl error=%d", ret);
+		return -EIO;
+	}
+
+	ret = H264EncSetCodingCtrl(data->encoder, &codingctrl_cfg);
+	if (ret != H264ENC_OK) {
+		LOG_ERR("H264EncSetCodingCtrl error=%d", ret);
+		return -EIO;
+	}
+
+	/* set bit rate configuration */
+	ret = H264EncGetRateCtrl(data->encoder, &ratectrl_cfg);
+	if (ret != H264ENC_OK) {
+		LOG_ERR("H264EncGetRateCtrl error=%d", ret);
+		return -EIO;
+	}
+
+	/* Constant bitrate */
+	ratectrl_cfg.pictureRc = 0;
+	ratectrl_cfg.mbRc = 0;
+	ratectrl_cfg.pictureSkip = 0;
+	ratectrl_cfg.hrd = 0;
+	ratectrl_cfg.qpHdr = VENC_DEFAULT_QP;
+	ratectrl_cfg.qpMin = ratectrl_cfg.qpHdr;
+	ratectrl_cfg.qpMax = ratectrl_cfg.qpHdr;
+
+	ret = H264EncSetRateCtrl(data->encoder, &ratectrl_cfg);
+	if (ret != H264ENC_OK) {
+		LOG_ERR("H264EncSetRateCtrl error=%d", ret);
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int encoder_start(struct stm32_venc_data *data, struct video_buffer *output)
+{
+	H264EncRet ret;
+	H264EncIn encIn = {0};
+	H264EncOut encOut = {0};
+
+	encIn.pOutBuf = (uint32_t *)output->buffer;
+	encIn.busOutBuf = (uint32_t)encIn.pOutBuf;
+	encIn.outBufSize = ROUND_UP(output->size, ALIGNMENT_INCR);
```

Comment:
Should this be capped to the `fmt->size` announced? As it is said to be the maximum size by `video.h`.

Which semantics would you wish for fmt->size? If it is the max size, then `fmt->max_size` would be fit? I will follow-up on https://github.com/zephyrproject-rtos/zephyr/pull/92884#discussion_r2409617631